_adr/83/index.adoc

```diff
@@ -0,0 +1,156 @@
+---
+num: 83
+title: Permission API
+status: "Accepted"
+authors:
+  - "Massimiliano Ziccardi"
+tags:
+  - "kafka"
+#applies_padrs: # What PADRs does this ADR apply?
+#applies_patterns: # What APs does this ADR apply?
+---
+## Context and Problem Statement
+Deciding whether a certain operation is allowed can be a complex matter and may require interactions with different components.
+
+Consumers of the Kas Fleet Manager APIs have no way of verifying in advance whether a particular request will be successful.
+The only way to do that, is to perform the same checks that Kas Fleet Manager will do while executing the action.
+
+This brings complexities and duplications. For example, just to check whether the creation of a new kafka is possible or not,
+currently UI and CLI have to perform at least the following checks:
+
+. Verify that the required terms and conditions have been accepted (this requires interaction with another service)
+. Make sure you have enough quota allocated. This requires:
+.. Having knowledge of the AMS endpoints and how to deal with them
+.. Having knowledge of the AMS terms used for designing the instance types (RHOSAKTrial for eval/developer and RHOSAK for standard)
+.. Enquire the quota assigned using the AMS endpoints, loop through the resources and check that enough quota is still available for product RHOSAK and resource RHOSAK
+.. If you have no quota assigned for STANDARD instances (RHOSAK), fallback to EVAL
+.. If you have quota for STANDARD (RHOSAK), but it is 0, fallback to EVAL (RHOSAKTrial)
+.. If you have quota for STANDARD (RHOSAK) and it is greater than 0 but all the granted quota is consumed, the creation should fail
+. Verify that at least one OCD Cluster has capacity for the instance you are going to create
+
+Another point of attention is that currently, any change in the flow requires an update to all the actors/consumers who are carrying out those checks.
+
+## Goals
+* API consumers can query a new KAS Fleet Manager Permission API to check if they can perform a certain action and receive a response with the outcome (allowed/denied) containing all the granted resources or, eventually, the missing requirements.
+
+## Non-goals
+
+## Current architecture
+* Kas Fleet Manager currently only allows the ‘try and see if it works’ approach: if what you are trying to do fails, Kas Fleet Manager gives you some details on went wrong, but you have no way of checking in advance if what you want to do is possible
+* API Consumers (like UI) are doing the required checkss into their codebase and every time a change occurs, both Kas Fleet Manager and UI have to be updated to reflect the change
+
+## Proposed architecture
+* We need to create a new API.
+The API will be authenticated and will allow the caller to check in advance whether he can perform a certain operation:
++
+```
+/can/{verb}/{resource}?deep=true
```

Comment:
Now is this API versioned?

Reply:
yeah, all our API are versioned. I removed the `/api/v1` prefix for clarity

Reply:
Thanks, I wasn't certain.